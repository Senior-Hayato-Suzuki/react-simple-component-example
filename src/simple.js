import React from "react";
import cn from "classnames";
import {css} from "glamor";
import omit from "lodash/fp/omit";
import pick from "lodash/fp/pick";
import pickBy from "lodash/fp/pickBy";
import deepmerge from "deepmerge";

const pickTruthty = pickBy(Boolean);

function getDisplayName(Component) {
    if (typeof Component === "string") {
        return Component;
    }

    return Component.displayName || Component.name || "Anonymous";
}

var compnentCounter = 0;

function simple(Component, styles, alts={}, options={}) {
    const num = compnentCounter++;
    const makeDebugClass = className => className
            ? className + "___" + num
            : null;

    const level = options._level || 0;

    if (Component._styleWrapped) {
        return simple(
            Component._styleWrapped.Component,
            deepmerge(Component._styleWrapped.styles, styles),
            deepmerge(Component._styleWrapped.alts, alts),
            {
                _level: level + 1,
            }
        );
    }

    const rules = {__base: css(styles)};

    const altProps = Object.keys(alts);


    for (let key in alts) {
        rules[key] = css(deepmerge(styles, alts[key]));
    }

    function StyleWrapped({className, ...otherProps}) {

        const passProps = omit(altProps, otherProps);
        const alt = Object.keys(pickTruthty(pick(altProps, otherProps)));


        if (alt.length > 1) {
            throw new Error("Too many alt props: " + alt.join(", "));
        }

        const rule = rules[alt[0] || "__base"];


        const props = {
            ...passProps,
            className: cn(makeDebugClass(StyleWrapped.displayName), makeDebugClass(alt[0]), className, String(rule)),
        };

        return <Component {...props} />;
    }

    StyleWrapped._styleWrapped = {
        Component,
        styles,
        alts,
    };

    StyleWrapped.displayName = "simple(" + getDisplayName(Component) + ")";

    StyleWrapped.create = el => simple(el, styles, alts);


    return StyleWrapped;
}


const viewStyles = {
    boxSizing: "border-box",
    position: "relative",
    display: "flex",
    flexDirection: "column",
    alignItems: "stretch",
    flexShrink: 0,
    alignContent: "flex-start",
    border: 0,
    margin: 0,
    padding: 0,
    minWidth: 0,
};

export const View = simple("div", viewStyles);
View.displayName = "View";


export {css};
export default simple;
