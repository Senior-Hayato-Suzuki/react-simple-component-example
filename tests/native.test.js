import React from "react";
// eslint-disable-next-line import/no-unresolved
import {View, StyleSheet} from "react-native";
import renderer from "react-test-renderer";

import simple from "../src/native";

jest.mock(
    "react-native",
    () => ({
        View: "View",
        StyleSheet: {
            create: jest.fn(s => s),
        },
    }),
    {virtual: true},
);

describe("native backend", () => {
    beforeEach(() => {
        StyleSheet.create.mockClear();
    });

    test("renders styles", () => {
        const MyView = simple(View, {
            padding: 10,
        });

        const tree = renderer.create(<MyView />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("is called with the style definition", () => {
        const MyView = simple(View, {
            padding: 10,
        });
        MyView; // eslint-disable-line

        expect(StyleSheet.create.mock.calls).toEqual([
            [
                {
                    base: {
                        padding: 10,
                    },
                },
            ],
        ]);
    });

    test("is called with the alt style definition", () => {
        const MyView = simple(
            View,
            {
                padding: 10,
            },
            {
                red: {
                    backgroundColor: "red",
                },
            },
        );

        expect(StyleSheet.create.mock.calls).toEqual([
            [{base: {padding: 10}}],
            [{base: {padding: 10}}],
            [{base: {backgroundColor: "red"}}],
        ]);

        const tree = renderer.create(<MyView red />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("is called with the extended styles", () => {
        const MyView = simple(View, {
            padding: 10,
        });

        const RoundButton = simple(MyView, {
            borderRadius: 10,
        });

        expect(StyleSheet.create.mock.calls).toEqual([
            [{base: {padding: 10}}],
            [{base: {padding: 10}}],
            [{base: {borderRadius: 10}}],
        ]);

        const tree = renderer.create(<RoundButton />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
