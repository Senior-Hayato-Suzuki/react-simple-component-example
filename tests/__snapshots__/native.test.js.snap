// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`native backend is called with the alt style definition 1`] = `
<View
  style={
    Array [
      Object {
        "padding": 10,
      },
      Object {
        "backgroundColor": "red",
      },
    ]
  }
/>
`;

exports[`native backend is called with the extended styles 1`] = `
<View
  style={
    Array [
      Object {
        "padding": 10,
      },
      Object {
        "borderRadius": 10,
      },
    ]
  }
/>
`;

exports[`native backend renders styles 1`] = `
<View
  style={
    Array [
      Object {
        "padding": 10,
      },
    ]
  }
/>
`;
